astropy/io/fits/tests/test_fsspec.py

```diff
@@ -0,0 +1,125 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+"""Can `astropy.io.fits.open` access (remote) data using the fsspec package?
+"""
+import numpy as np
+import pytest
+from numpy.testing import assert_allclose, assert_array_equal
+
+from astropy.io import fits
+from astropy.nddata import Cutout2D
+from astropy.utils.compat.optional_deps import HAS_FSSPEC, HAS_S3FS  # noqa
+from astropy.utils.data import conf, get_pkg_data_filename
+
+if HAS_FSSPEC:
+    import fsspec
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_local():
+    """Can we use fsspec to read a local file?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    with fits.open(fn) as hdulist_classic:
+        with fits.open(fn, use_fsspec=True) as hdulist_fsspec:
+            assert_array_equal(hdulist_classic[2].data, hdulist_fsspec[2].data)
+            assert_array_equal(hdulist_classic[2].section[3:5], hdulist_fsspec[2].section[3:5])
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_local_write(tmpdir):
+    """Can we write to a local file that was opened using fsspec?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    fn_tmp = tmpdir / "tmp.fits"
+    with fits.open(fn, use_fsspec=True) as hdul:
+        # writing to a section is never allowed
+        with pytest.raises(TypeError):
+            hdul[1].section[0, 0] = -999
+        # however writing to .data should work
+        hdul[1].data[2, 3] = -999
+        assert hdul[1].data[2,3] == -999
+        hdul.writeto(fn_tmp)
+
+    # Is the new value present when we re-open the file?
+    with fits.open(fn_tmp) as hdul:
+        assert hdul[1].data[2,3] == -999
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_cutout2d():
+    """Does Cutout2D work with data loaded lazily using fsspec and .section?"""
+    fn = get_pkg_data_filename('data/test0.fits')
+    with fits.open(fn, use_fsspec=True) as hdul:
+        position = (10, 20)
+        size = (2, 3)
+        cutout1 = Cutout2D(hdul[1].data, position, size)
+        cutout2 = Cutout2D(hdul[1].section, position, size)
+        assert_allclose(cutout1.data, cutout2.data)
+
+
+@pytest.mark.skipif("not HAS_FSSPEC")
+def test_fsspec_compressed():
+    """Does fsspec support compressed data correctly?"""
+    # comp.fits[1] is a compressed image with shape (440, 300)
+    fn = get_pkg_data_filename('data/comp.fits')
+    with fits.open(fn, use_fsspec=True) as hdul:
+        # The .data attribute should work as normal
+        assert hdul[1].data[0,0] == 7
+        # However the .section attribute does not support compressed data
+        with pytest.raises(AttributeError) as excinfo:
+            hdul[1].section[1,2]
+        assert "'CompImageHDU' object has no attribute 'section'" in str(excinfo.value)
```

Comment:
Why not use `match` here? (Please make sure the suggestion works locally because this uses regex.)

```suggestion
        with pytest.raises(AttributeError, match="'CompImageHDU' object has no attribute 'section'") as excinfo:
            hdul[1].section[1,2]
```